serde-generate/src/indent.rs

```diff
@@ -0,0 +1,135 @@
+use std::io::{Result, Write};
+
+#[derive(Clone, Copy)]
+pub enum Indent {
+    Tab,
+    Space(usize),
+}
+
+pub struct IndentedWriter<T> {
+    w: T,
+    indent_level: usize,
+    indent: Indent,
+    at_begining_of_line: bool,
+}
+
+impl<T> IndentedWriter<T> {
+    pub fn new(w: T, indent: Indent) -> Self {
+        Self {
+            w,
+            indent_level: 0,
+            indent,
+            at_begining_of_line: true,
+        }
+    }
+
+    pub fn level(&self) -> usize {
+        self.indent_level
+    }
+
+    pub fn set_level(&mut self, level: usize) {
+        self.indent_level = level
+    }
+
+    pub fn inc_level(&mut self) {
+        self.inc_level_by(1);
+    }
+
+    pub fn dec_level(&mut self) {
+        self.dec_level_by(1);
+    }
+
+    pub fn inc_level_by(&mut self, n: usize) {
+        self.indent_level = self.indent_level.saturating_add(n);
+    }
+
+    pub fn dec_level_by(&mut self, n: usize) {
+        self.indent_level = self.indent_level.saturating_sub(n);
+    }
```

Comment:
I think we only need `inc_level` (aka "indent" ?) and `dec_level` ("unindent"?) in practice. Also, it would be really nice to find a trick to remove `dec_level` and use RAII (scoping) instead.

Reply:
In practice that's true. of course right now I'm using `set_level` since it hasn't been piped through everywhere yet. Perhaps once we get used to using this we play with RAII and removing these extra methods?

Reply:
We can certainly experiment with one file (say java.rs) but I would rather refactor the rest of the files only once, with the final (hopefully idiot-proof) API.

Reply:
@bmwill I tried some refactoring on my side: https://github.com/facebookincubator/serde-reflection/pull/33. RAII is not playing too well with `&mut out`. Also, all our errors are final (uncaught) so maybe we just don't need this.

Reply:
when you say "this" are you refering to RAII?

Reply:
Yes. I agree with you that it's simpler to try RAII at the very end.